### PR TITLE
Refactor damage resist

### PR DESCRIPTION
**Pull request by Kribylet**
#### Summary
Bugfixes "refactor item::*_resist, fix bugs related to clothing mods and outfit bodypart coverage"

#### Purpose of change

When looking over the armor calculations a few days ago to try and resolve an issue with electric damage causing item damage, I noticed there seemed to be number of bugs introduced whenever changes happened, in particular due to the large amount of code duplication.

This PR aims to refactor armor and item damage resistance calculations to reduce code duplication (currently duplicated 2x<#damage types>).

While doing the refactoring I noticed that several unrelated damage types seemed to benefit from clothing mods that boosted bash resistance, presumably due to code chunks being copied and pasted and eyes glazing over trying to read all the near identical function calls.

Another oversight seems to have been found in outfit::damage_resist(...) which did not propagate which bodypart was covered by an item for armor portion calculations.

#### Describe the solution

<s>The damage_type enum class was moved from damage.h to enums.h. Creature.h, damage.h and item.h were made to rely on the enums.h definition instead of forward-declaring damage_type.</s> The enums were reordered to allow for a single comparison when checking whether incoming damage was capable of damaging items. This adds implicit meaning to the damage_type order of damage types.

The functions material_type::*_resist were deduplicated by introducing a std::map container to hold resistance values. A single function material_type::resist( damage_type ) is used to replace the previous functionality. The initialization of the map values is done in the material_type constructor.

The function overload get_thickness( sub_bodypart_id& ) was introduced to generalize bodypart_id and sub_bodypart_id handling in the item::resist_* calculations.

A new function item::get_clothing_mod_val_for_damage_type( damage_type ) was introduced to retrieve clothing mod bonuses based on the damage type being considered.

With these changes in place, all the old item::*_resist calls were refactored to a new function, item::resist. The new function uses template instantiation and the new item::get_thickness overload to handle calls for both bodypart_id and sub_bodypart_id, preferring to stub bodypart_id if no bodypart is specified. It extracts the required information from the bodyparts if any and forwards that to a private function, item::_resist. That function in turn handles the old armor and item resist calculations as before, and forwards to item::_environmental_resist for fire and acid damage type handling. 

#### Describe alternatives you've considered

Leaving enum class damage_type in damage.h; not possible if material.h/material.cpp should be able to do dynamic retrieval of damage type entries for items, something that was necessary for the deduplication.

Instead of adding implicit meaning to the order of damage_type enums; messing with damage_type enums to add a standoff entry like "CAN_DAMAGE_ITEMS", or specifying a static integer offset for damage types that can damage items. Both introduce weirdness such as changing array size when trying to rely on static_cast<int>, or special values which must be ignored when iterating over resistance types.

#### Testing

Ran around for a bit and got beat up, had armor absorb strikes so the avatar didn't take damage, had armor absorb strikes and not get damaged, had armor absorb strikes and get damaged. Was struck by electric damage which did not damage armor.

### CODE REVIEW

**Comment by bombasticSlacks** [comment]
Awesome I was gonna do this refactor at some point. There's a lot of nitty gritty so I'll have to take some time to review. I'll skim now but I might need to do some testing and stuff.

**Comment by bombasticSlacks** [comment]
Okay so I'm not the most confident in reviewing your fancy template-ing. I think the public function should be put in the CPP file instead of the header (though I'm not sure what wrapping would be required for that). Overall I think this looks great, for context the reason we were holding off on cleaning this up was that the plan is to at some point JSONize damage types so mods can extend the available damage types (this shouldn't be done rn). However this clean up is great since the indexed vector you put them in could easily be swapped to a map of IDs once that extension happens. I'm gonna ping @anothersimulacrum to give this a scan since they were also looking at this stuff (and as far as I know they are better with nitty gritty C++ over me). 

However other than my misgivings about the amount of logic resist puts in a header I approve of this 👍

**Comment by anothersimulacrum** [comment]
More full review coming, but just a note:
> The damage_type enum class was moved from damage.h to enums.h. Creature.h, damage.h and item.h were made to rely on the enums.h definition instead of forward-declaring damage_type.

This shouldn't be done unless it's necessary (and it's not for at least one file, as far as I saw). The more files you include a header in, the more files must be recompiled when you edit that file, and when they're headers, that chains. `item.h` and `character.h` are huge headers, every file you include in them become both quite painful to edit, and then slows compile times by including that header. Forward declare unless you cannot.

**Comment by Kribylet** [comment]
I'm going to be a bit slow this week due to traveling to be with family, but I wanted to put some brief thoughts regarding the feedback so far. I think I can roll back the header changes and avoid that bit entirely by just moving logic regarding specific damage types into new helper functions instead. 

The public template function being in the header is an issue with how template instantiation has to happen in the translation unit that tries to use and deduce the types of the function declared by the template, where it does not have access to the definitions outside of the header. For example, when material_type in material.cpp uses obj.resist(...), the entire template definition must be visible from the item header in order to perform instantiation. One way to weasel out of defining the template function in the header is to include another header file to store the definition (that included header would then be visible to any translation unit that has the initial header), but that feels unsatisfactory to me. All that being said, a lot of the game logic can be moved to the private function while keeping only the sub_bodypart_id and bodypart_id info extraction in the header. The only downside of this is that some overhead is introduced even for calls that previously exited nearly instantly (is_null(), damage_type::NONE, to_self && dmg_type_cant_damage_items etc).

File: src/item.h [hunk 1]
```diff
@@ -1184,56 +1183,99 @@ class item : public visitable
...
+         *                     against each item material individually. For environmental type damage such as
+         *                     fire and acid, this value can be used to allow hypothetical calculations for
+         *                     environmental protection items such as gas masks.
+         */
+        /*@{*/
+
+        template<typename bodypart_target = bodypart_id>
+        typename std::enable_if_t < std::is_same<bodypart_target, bodypart_id>::value ||
+        std::is_same<bodypart_target, sub_bodypart_id>::value, float >
+        resist( const damage_type dmg_type, bool to_self = false,
```
Comment (andrei8l) [inline src/item.h:1227]: `std::enable_if_t<>` is just noise if you're not specializing this template further. 

You can move the code to the cpp file then use `extern template` in the header and manually instantiate in the cpp (see for example `Character::can_lift<>()`).

File: src/item.h [hunk 2]
```diff
@@ -1184,56 +1183,99 @@ class item : public visitable
...
+            const bool bp_null = bp == bodypart_target();
+            const std::vector<const part_material *> &armor_mats = armor_made_of( bp );
+            const float avg_thickness = bp_null ? get_thickness() : get_thickness( bp );
+            return _resist( dmg_type, to_self, resist_value, bp_null, armor_mats, avg_thickness );
+        }
+
+    private:
+        float _resist( damage_type dmg_type, bool to_self = false, int resist_value = 0,
+                       bool bp_null = true,
+                       const std::vector<const part_material *> &armor_mats = std::vector<const part_material *>(),
```
Comment (andrei8l) [inline src/item.h:1259]:
```suggestion
                       const std::vector<const part_material *> &armor_mats = {},
```
same thing for `_environmental_resist` below

File: src/material.cpp [hunk 3]
```diff
@@ -63,6 +63,10 @@ material_type::material_type() :
     _bash_dmg_verb( to_translation( "damages" ) ),
     _cut_dmg_verb( to_translation( "damages" ) )
 {
+    // negative integers means susceptibility
+    for( size_t i = 0; i < static_cast<int>( damage_type::NUM ); i++ ) {
```
Comment (andrei8l) [inline src/material.cpp:67]:
```suggestion
    for( int i = 0; i < static_cast<int>( damage_type::NUM ); i++ ) {
```

File: src/material.h [hunk 4]
```diff
@@ -79,14 +80,7 @@ class material_type
...
         itype_id _repaired_with = itype_id( "null" ); // this material can be repaired with this item
-        float _bash_resist = 0.0f;                         // negative integers means susceptibility
-        float _cut_resist = 0.0f;
-        float _acid_resist = 0.0f;
-        float _elec_resist = 0.0f;
-        float _fire_resist = 0.0f;
-        float _bullet_resist = 0.0f;
-        float _biologic_resist = 0.0f;
-        float _cold_resist = 0.0f;
+        std::map<damage_type, float> _resistances{};
```
Comment (andrei8l) [inline src/material.h:83]:
```suggestion
        std::map<damage_type, float> _resistances;
```

**Comment by LyleSY** [comment]
Looks like something's gone wonky here. Maybe rebase? Guessing you didn't mean to make so many big changes to DinoMod here

**Comment by Kribylet** [comment]
> Looks like something's gone wonky here. Maybe rebase? Guessing you didn't mean to make so many big changes to DinoMod here

Very correct! I ruined the remote version with my local after trying to get back into working on this. Now I'm completely restructuring this PR into a single commit to reduce the rebase nightmare it has become, and to clean up the rebase. Hopefully it won't be too hard to review given the discussion history.

File: src/item.h (hunk 1 above)
Comment (Kribylet) [reply src/item.h]: This change should be incorporated now.

**Comment by Kribylet** [comment]
I've re-written this PR based on previous commits and feedback and brought it up to speed with the main branch, while squishing it down to a single commit. I did this to reduce troublesome rebase issues that popped up from splitting the PR into smaller-scope (though non-functional) intermediate commits.

While the PR is less easy to review this way, it should hopefully be much easier to rebase it once it passes muster sometime post feature thaw.

I re-ran the basic tests from before:
>Ran around for a bit and got beat up, had armor absorb strikes so the avatar didn't take damage, had armor absorb strikes and not get damaged, had armor absorb strikes and get damaged. Was struck by electric damage which did not damage armor.

As extra icing on the cake I also tried wearing complete electric insulation and was immune to electric damage.

**Comment by bombasticSlacks** [comment]
Looks great, thanks for your work on this, sorry it took an eternity to review.

**Pull request merged**